### PR TITLE
Lints and Reformatting

### PR DESCRIPTION
**Pull request by fosskers**
This PR clears out a number of warnings, and improves code formatting in a number of places. Further, it solves a critical bug found in `group` and `groupBy`.

### CODE REVIEW

File: Data/ByteString/Streaming.hs
Comment (fosskers) [inline Data/ByteString/Streaming.hs]: This function was unused.

File: Data/ByteString/Streaming.hs
Comment (fosskers) [inline Data/ByteString/Streaming.hs]: This branch of the pattern match was missing. I'm not exactly comfortable with an `error` here, but I'm unsure what was intended by the original code.

File: Data/ByteString/Streaming.hs
Comment (fosskers) [inline Data/ByteString/Streaming.hs]: Same as above.

**Comment by fosskers** [comment]
Not quite out of the frying pan yet. Not sure why it's failing for just those two GHC versions.

**Comment by fosskers** [comment]
Rebased on top of `master`.

**Comment by fosskers** [comment]
Alright, it's failing at the same place, which means this is 100% my fault.

**Comment by fosskers** [comment]
Ok, seems good-to-go, just wondering about that suspicious pattern match branch above.

File: Data/ByteString/Streaming.hs
Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: It seems you've found a bug.  That case is NOT impossible.  Here's an example that crashes `group`:
```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where
import qualified Data.ByteString.Streaming as Q
import qualified Data.ByteString.Streaming.Internal as Q
import qualified Streaming as S
import qualified System.IO as IO

main :: IO ()
main = do
    let x = Q.Chunk "f" (Q.Go (pure $ Q.Empty ()))
        y = Q.group x
    S.mapsM_ (Q.hPut IO.stdout) y
```
Thus I get:
```
[1 of 3] Compiling Data.ByteString.Streaming.Internal ( Data/ByteString/Streaming/Internal.hs, interpreted )
[2 of 3] Compiling Data.ByteString.Streaming ( Data/ByteString/Streaming.hs, interpreted )
m[3 of 3] Compiling Main             ( foo.hs, interpreted )
Ok, three modules loaded.
λ> main
*** Exception: Data/ByteString/Streaming.hs:(1209,3)-(1220,78): Non-exhaustive patterns in function to
```

Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: I see the same problem in `groupBy`. Neither have _ever_ been correct. These look like they're not much used, and so poorly tested...

File: Data/ByteString/Streaming/Internal.hs
Comment (vdukhovni) [inline Data/ByteString/Streaming/Internal.hs]: The `bytesting` library is about to remove the long deprecated `inlinePerformIO` as of `0.11` you should use either `accursedUnutterablePerformIO` or (better yet) `unsafeDupablePerformIO` (provided GHC 7.0 is not supported, the latter is only available with GHC 7.2 and up).  The release changelog PR in `bytestring`.

Comment (vdukhovni) [reply Data/ByteString/Streaming/Internal.hs]: [See] (https://github.com/haskell/bytestring/commit/444c475afc34d78d89bae7449c9109f6ee617cae)
By the way it seems this library uses both of the legacy functions, but given that GHC 7.10 is the oldest supported (or at least oldest tested) swapping both for `unsafeDupablePerformIO` seems right to me.

Cc: @bodigrim

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Nice to see some long-standing issues getting a look.  Almost there...

File: Data/ByteString/Streaming.hs
Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: Here's another sample program:
```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where
import qualified Data.ByteString.Streaming as Q
import qualified Data.ByteString.Streaming.Char8 as Q8
import qualified Data.ByteString.Streaming.Internal as Q
import qualified Streaming as S
import qualified System.IO as IO

main :: IO ()
main = do
    let x = Q.Chunk "a"
          $ Q.Go
          $ pure
          $ Q.Chunk "a"
          $ Q.Go
          $ pure
          $ Q.Chunk "bbc"
          $ Q.Go
          $ pure
          $ Q.Chunk "c"
          $ Q.Go
          $ pure
          $ Q.Empty ()
        y = Q.group x
    S.mapsM_ Q8.putStrLn y
```
and with the patch below it produces the expected:
```
[1 of 4] Compiling Data.ByteString.Streaming.Internal ( Data/ByteString/Streaming/Internal.hs, interpreted )
[2 of 4] Compiling Data.ByteString.Streaming ( Data/ByteString/Streaming.hs, interpreted )
[3 of 4] Compiling Data.ByteString.Streaming.Char8 ( Data/ByteString/Streaming/Char8.hs, interpreted )
[4 of 4] Compiling Main             ( foo.hs, interpreted )
Ok, four modules loaded.
λ> main
aa
bb
cc
```
Proposed patch:
```haskell
diff --git a/Data/ByteString/Streaming.hs b/Data/ByteString/Streaming.hs
index b696c24..be77b51 100644
--- a/Data/ByteString/Streaming.hs
+++ b/Data/ByteString/Streaming.hs
@@ -1206,17 +1206,13 @@ group = go
     | otherwise        = Step $ to [S.unsafeTake 1 c] (S.unsafeHead c)
                                      (Chunk (S.unsafeTail c) cs)
 
-  to acc !_ (Empty r)        = revNonEmptyChunks
-                                     acc
-                                     (Empty (Return r))
+  to acc !_ (Empty r)    = revNonEmptyChunks acc (Empty (Return r))
+  to acc !w (Go m)       = Go $ to acc w <$> m
   to acc !w (Chunk c cs) =
     case findIndexOrEnd (/= w) c of
-      0                    -> revNonEmptyChunks
-                                    acc
-                                    (Empty (go (Chunk c cs)))
+      0                    -> revNonEmptyChunks acc (Empty (go (Chunk c cs)))
       n | n == S.length c  -> to (S.unsafeTake n c : acc) w cs
-        | otherwise        -> revNonEmptyChunks
-                                    (S.unsafeTake n c : acc)
+        | otherwise        -> revNonEmptyChunks (S.unsafeTake n c : acc)
                                     (Empty (go (Chunk (S.unsafeDrop n c) cs)))
 
 {-# INLINABLE group #-}
```

That is, I think the missing branch should be:
```haskell
    to acc !w (Go m)       = Go $ to acc w <$> m
```

**Comment by vdukhovni** [comment]
Since clearly nobody has used `group` or `groupBy`, and it is not clear why they're needed, they could be dropped rather than fixed?  What use-case justifies their existence?  Of course, they could be fixed, and deprecated first, then see if anyone objects to the deprecation.

File: Data/ByteString/Streaming.hs
Comment (fosskers) [reply Data/ByteString/Streaming.hs]: It seems easy to trigger the bug when manually using the `Streaming.ByteString` constructors, but what about when this is coming from a string literal or a file? What concrete string-based example would trigger this?

**Comment by fosskers** [comment]
Thanks a ton for catching that bug, I'm glad I left the PR open. At the very least I should add a test case for those to the unit tests. Indeed very strange that these were never encountered in the wild before.

File: Data/ByteString/Streaming.hs
Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: Well, any streaming bytestring that has a multi-chunk "skeleton" requiring monadic steps should do.  So if you take a file larger than 32k bytes, this should break trying to determine whether the last byte of the first chunk is matched by the first byte of the next chunk (which requires an IO operation to bring it into memory).

Comment (fosskers) [reply Data/ByteString/Streaming.hs]: I'll produce a test file and add a unit test.

**Comment by vdukhovni** [comment]
> Thanks a ton for catching that bug, I'm glad I left the PR open. At the very least I should add a test case for those to the unit tests. Indeed very strange that these were never encountered in the wild before.

Good call.  I stumbled into this somewhat at random, just looking to see whether you've cut a release... Idle curiosity.

File: Data/ByteString/Streaming.hs
Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: If you open the file with a smaller (non-default) I/O buffer size, the test file can be smaller...

**Review by vdukhovni — COMMENTED** [review]
And I hope you noticed the comment about `inlinePerformIO`...

File: Data/ByteString/Streaming.hs
Comment (fosskers) [reply Data/ByteString/Streaming.hs]: This makes it fail:
```haskell
groupCrash :: Assertion
groupCrash = do
  res <- runResourceT . S.length_ . SM.mapsM SBS8.length . SBS8.group @(ResourceT IO) $ SBS8.readFile "tests/groupBy.txt"
  res @?= 39925
```

**Comment by fosskers** [comment]
I did notice the line regarding the inline IO, I will get to that next.

File: Data/ByteString/Streaming.hs
Comment (vdukhovni) [reply Data/ByteString/Streaming.hs]: Did you mean for the `S.length_` to be `S.sum_`?

Comment (fosskers) [reply Data/ByteString/Streaming.hs]: It now is. When I said "fail" I meant "crash" due to the bad pattern match. It's all fixed now! :)

File: streaming-bytestring.cabal
Comment (vdukhovni) [inline streaming-bytestring.cabal]: You probably noticed that the `resourcet` version bound looks too high for some of the older LTS snapshots in the test matrix...

Comment (fosskers) [reply streaming-bytestring.cabal]: Yes, I was waiting for CI to fail and tell me :)

**Review by vdukhovni — APPROVED** [review]
LGTM, modulo CI tweaks, which you can surely sort out...

**Comment by chessai** [comment]
Please use accursedUnutterablePerformIO instead of unsafeDupablePerformIO,
they are not the same and the former performs better

**Review by vdukhovni — CHANGES_REQUESTED** [review]
While you're at it, you may as well also replace the `accursedUnutterablePeformIO` calls too.

**Comment by vdukhovni** [comment]
> Please use accursedUnutterablePerformIO instead of unsafeDupablePerformIO, they are not the same and the former performs better

Are you sure there's a significant difference?  I didn't see it in any tests I ran?

**Comment by vdukhovni** [comment]
> > Please use accursedUnutterablePerformIO instead of unsafeDupablePerformIO, they are not the same and the former performs better
> 
> Are you sure there's a significant difference? I didn't see it in any tests I ran?

Whichever is used, it should be the same function in all cases, rather than an ad-hoc mixture.

**Comment by chessai** [comment]
There have been benchmarks performed in the past that indicated a
significant difference. These are findable somewhere on the libraries
mailing list, though I can't remember where. I think the significant
difference is in inlining. unsafeDupablePerformIO uses runRW#, for which
inlining is unsound, so GHC always inlines it after inlining everything
else. This is just a hunch. For now I think we should stop using
inlinePerformIO and stick with accursedUnutterablePerformIO.

**Comment by fosskers** [comment]
That should do it. I'll merge later if there are no objections.

**Comment by vdukhovni** [comment]
I notice that many instances of `liftM` changed to `fmap`. I am curious whether there's more here than just notational difference?

**Comment by fosskers** [comment]
It's possible that the original author intended something specific by `liftM`. Typewise, they're equivalent, hence the drop-in replacement.

**Comment by vdukhovni** [comment]
The `IO` instance of `Functor` and `Monad` are below, along with the definition of `liftM`:

```haskell
instance  Functor IO where
   fmap f x = x >>= (pure . f)

instance  Monad IO  where
    {-# INLINE (>>)   #-}
    {-# INLINE (>>=)  #-}
    (>>)      = (*>)
    (>>=)     = bindIO

liftM   :: (Monad m) => (a1 -> r) -> m a1 -> m r
liftM f m1              = do { x1 <- m1; return (f x1) }
```

**Comment by fosskers** [comment]
Likewise, here's a use site:
```haskell
unpack ::  Monad m => ByteString m r ->  Stream (Of Char) m r
unpack bs = case bs of
    Empty r    -> Return r
    Go m       -> Effect (fmap unpack m)
    Chunk c cs -> unpackAppendCharsLazy c (unpack cs)
```
Should we be concerned about something to do with laziness here?

**Comment by vdukhovni** [comment]
> Likewise, here's a use site:
> 
> ```haskell
> unpack ::  Monad m => ByteString m r ->  Stream (Of Char) m r
> unpack bs = case bs of
>     Empty r    -> Return r
>     Go m       -> Effect (fmap unpack m)
>     Chunk c cs -> unpackAppendCharsLazy c (unpack cs)
> ```
> 
> Should we be concerned about something to do with laziness here?

The original had `liftM` rather than `fmap`, it is not obvious whether there's any difference. What did hlint have to say?

**Comment by fosskers** [comment]
```
Char8.hs:203:27-31: Warning: Use fmap
Found:
  liftM
Perhaps:
  fmap
```

**Comment by fosskers** [comment]
I'm going to pose this question to the [Haskell Discourse](https://discourse.haskell.org/).

**Comment by fosskers** [comment]
https://discourse.haskell.org/t/use-of-liftm-over-fmap/1412

**Comment by vdukhovni** [comment]
> ```
> Char8.hs:203:27-31: Warning: Use fmap
> Found:
>   liftM
> Perhaps:
>   fmap
> ```

Right and of course we could also write: `Effect (m >>= return . unpack)` or eta expand to `Effect (m >>= \x -> return (f x)`.  Or even `Effect (m >>= \x -> return $! f x)`.  I don't yet have a solid feel for which of these is best or even different...

**Comment by vdukhovni** [comment]
I think we can safely go with `fmap`.

When inlined (caller and function using one of the idioms in the same module) I could not find any difference between `fmap f m`, `liftM f m`, `f <$> m` and  `m >>= return . f`, they all produced identical code.

When not inlined, the code generated for the`liftM` and `>>=` variants differed slightly from the `fmap` and `<$>` variants, but performance was not noticeably affected, though there was perhaps a tiny bias in favour of the first two. (3.108s runtime for the first two vs 3.118s runtime for the second two).

Only `m >>= \x -> return $! f x` produced different code when inlined, but in naïve tests it offered no advantage either when inlined or not, seemed to be about 1% slower in both cases.

Now admittedly, I was measuring with 32kB IO chunks where the `effect` processing is a rather modest portion of the cost, most of the time is spent on the content, not the I/O.  We could perhaps look into measuring costs when the content processing is lighter than with `group` and the I/O cost might be easier to spot.

Mostly I think what matters is whether a given function is inlined and specialisation gets to replace abstract bind operations with the particular monad at hand.  This is likely more significant than how we write `fmap`.

Functions too large for inlining, but that are mostly used with `m == IO` could have `SPECIALISE` pragmas that make the `IO` case efficient. Assuming the function is widely used, and not something barely or not at all useful like `group`.

**Comment by fosskers** [comment]
Thanks for all your analysis on this. Given that nobody else bit on the Discourse topic, let's move forward with `fmap`. This is a good lesson for future development too: namely that `hlint` can be trusted (I do trust Neil Mitchell's work in general).

**Comment by fosskers** [comment]
Testing out the new `github-cli` tool...

**Pull request merged**
